bcgov/biohubbc
# BHBC-1796-1810 : bug fix - xlsx observation upload not completing

# Overview

## Links to Jira tickets

https://quartech.atlassian.net/browse/BHBC-1796
https://quartech.atlassian.net/browse/BHBC-1810


Notable changes:
- transform xlsx step is faster
- new endpoint for the API to manage validate, transform and scrape


## Dev notes:
### The problem:

the UTM values should not have decimal points

There are three consecutive calls made from the app to process a submission: validate, transform, scrape.  The weak point was that since the file had a lot of rows, the last step of the transform that removes duplicates take a long time.  And when the API takes longer than 30 seconds to respond, the APP fails.

### The fix:
1. check the UTM values in the input file.  Down the road, this would be a good one to create a validation for. the format of the UTM values

2. Re-wrote the function that removes duplicates so that it is much faster

3. Introduced a new end-point xlsx/process that manages all the steps: validate, transform, and scrape.  The front-end now just makes one call.

4. The workflow for processing a submission now sends a success response back to the APP once the validation step is complete, and continues to process the submission in the background.  This way the APP doesn’t hang waiting.  This also means that you won’t see the points on the map until the processing is finished.


### To test:

Upload a file

If the file is large, give it a minute before refreshing the screen and you will see the points on the map.

Use this file to test: 
[modified_Moose_Aerial_StratifiedRandomBlock_Composition_Recruitment_Survey_2.5_Teslin2020DP.xlsx](https://github.com/bcgov/biohubbc/files/8920326/modified_Moose_Aerial_StratifiedRandomBlock_Composition_Recruitment_Survey_2.5_Teslin2020DP.xlsx)
